pkg/metadata/provider_azure_ovf.go

```diff
@@ -0,0 +1,171 @@
+package main
+
+import (
+	"encoding/base64"
+	"encoding/xml"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"path"
+	"path/filepath"
+	"syscall"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// ProviderAzureOVF extracts user data from ovf-env.xml. The file is on an Azure
+// attached DVD containing the user data, encoded as base64.
+// Inspired by:
+//  - WALinuxAgent:
+//	- https://github.com/Azure/WALinuxAgent
+//  - cloud-init Azure Datasource docs:
+//    https://cloudinit.readthedocs.io/en/latest/topics/datasources/azure.html
+type ProviderAzureOVF struct {
+	client     *http.Client
+	mountPoint string
+}
+
+// OVF XML model
+type OVF struct {
+	HostName       string `xml:"ProvisioningSection>LinuxProvisioningConfigurationSet>HostName"`
+	UserDataBase64 string `xml:"ProvisioningSection>LinuxProvisioningConfigurationSet>CustomData"`
+}
+
+// NewAzureOVF factory
+func NewAzureOVF(client *http.Client) *ProviderAzureOVF {
+	mountPoint, err := ioutil.TempDir("", "cdrom")
+	if err != nil {
+		panic(fmt.Sprintf("Azure-OVF: Creating temp mount dir failed: %s", err))
+	}
+	return &ProviderAzureOVF{
+		mountPoint: mountPoint,
+		client:     client,
+	}
+}
+
+func (p *ProviderAzureOVF) String() string {
+	return "Azure-OVF"
+}
+
+// Probe returns true if DVD is successfully mounted, false otherwise
+func (p *ProviderAzureOVF) Probe() bool {
+	if err := retry(6, 5*time.Second, p.mount); err != nil {
```

Comment:
Is this needed? Does the virtual CD take same time to appear??

Reply:
Yes, I used  the WALinuxAgent's osutil `default.py` implementation as reference, see:

https://github.com/Azure/WALinuxAgent/blob/v2.2.52/azurelinuxagent/common/osutil/default.py#L579-L618

On my test system `ls /tmp | grep cdrom` returns:

``` text
cdrom003897275
cdrom310295156
cdrom860156597
```

This would mean that two attempts failed. I thought about how nice it'd be to have regression tests for metadata packages. But how? Some thoughts (sorry, if I'm going offtopic):

* The tests would need to be labeled and not run by default. Not every host can run tests for all platforms
* For Azure you'd have to provision a VM / many VMs with the metadata package and verify its/their state after. Is that approach even feasible? Any inputs?

The impact of the `retry()` method is substantial. I also use it inside `ProviderAzureIMDS`. So an image that uses neither `AzureProviderIMDS` nor `AzureProviderOVF` would probe 30 seconds for each provider, amounting to 60 seconds of probing. So worst case, when `metadata` is run without args, this would amount to 60 seconds of just probing because of Azure providers.